ports/esp32/uart.c

```diff
@@ -28,42 +28,130 @@
 
 #include <stdio.h>
 
+#include "uart.h"
 #include "driver/uart.h"
 #include "soc/uart_periph.h"
 
 #include "py/runtime.h"
 #include "py/mphal.h"
 
-STATIC void uart_irq_handler(void *arg);
+static QueueHandle_t uart0_queue;
+
+static void uart_event_task(void *pvParameters);
+
 
 void uart_init(void) {
-    uart_isr_handle_t handle;
-    uart_isr_register(UART_NUM_0, uart_irq_handler, NULL, ESP_INTR_FLAG_LOWMED | ESP_INTR_FLAG_IRAM, &handle);
-    uart_enable_rx_intr(UART_NUM_0);
+    uart_config_t uartcfg = {
+        .baud_rate = 115200,
+        .data_bits = UART_DATA_8_BITS,
+        .parity = UART_PARITY_DISABLE,
+        .stop_bits = UART_STOP_BITS_1,
+        .flow_ctrl = UART_HW_FLOWCTRL_DISABLE,
+        .rx_flow_ctrl_thresh = 0
+    };
+    uart_param_config(UART_NUM_0, &uartcfg);
+
+    const uint32_t rxbuf = 260;
+    const uint32_t txbuf = 260;
+
+    uart_driver_install(UART_NUM_0, rxbuf, txbuf, 20, &uart0_queue, 0);
+
+    uart_repl_set_pattern_detect(mp_interrupt_char);
+
+    //Create a task to handler UART event from ISR
+    xTaskCreate(uart_event_task, "uart_event_task", 2048, NULL, 12, NULL);
+
+}
+
+void uart_repl_set_pattern_detect(const char pattern) {
+    //Set uart pattern detect function.
+    uart_enable_pattern_det_baud_intr(UART_NUM_0, pattern, 1, 9, 0, 0);
```

Comment:
Looks like this was introduced in IDF 4.1.  So it needs:
```
#if ESP_IDF_VERSION >= ESP_IDF_VERSION_VAL(4, 1, 0)
uart_enable_pattern_det_intr(UART_NUM_0, pattern, 1, 9, 0, 0);
#else
uart_enable_pattern_det_baud_intr(UART_NUM_0, pattern, 1, 9, 0, 0);
#endif
```

Reply:
Other way round, but yes good catch. I've just pushed that change, that might make the CI happy.